docs/RFCS/20210604_distributed_token_bucket.md

```diff
@@ -0,0 +1,416 @@
+dual-purpose, but things would become messy.
+
+Another possible variation would be to have a single `(tenant_id, op_id)`
+primary index and *also* store the current state and sequence number in
+`(tenant_id, '00000000-0000-0000-0000-000000000000')`. That would not require a
+secondary index, but would generate larger KVs as we store all fields twice.
+
+## KV Connector API
+
+The KV Connector API is used by the SQL pods to make requests to the distributed
+bucket. This API is implemented on the host cluster side by performing a
+transaction on `tenant_usage`.
+
+```go
+type TenantTokenBucketProvider interface {
+   Request(
+     ctx context.Context,
+     uuid util.UUID,
+     prevShareValue float64,
+     prevShareTimestamp time.Time,
+     tokensRequested float64,
+     ruUsage float64,
+     readRequests int64,
+     readBytes int64,
+     writeRequests int64,
+     writeBytes int64,
+     totalPodCPUSeconds float64,
+   ) (
+     updateTimestamp time.Time,
+     grantedTokens float64,
+     grantedTimestamp time.Time,
+     maxBurstTokens float64
+   )
+```
+
+
+## Configuration API
+
+The limits for a tenant are periodically reconfigured. The proposed interface is
+via an internal SQL function; we can easily provide other endpoints as
+necessary.
+
+```sql
+SELECT crdb_internal.update_tenant_resource_limits(
+  tenant_id,                    -- INT
+  operation_uuid,               -- UUID
+  as_of,                        -- TIMESTAMP
+  as_of_consumed_request_units, -- FLOAT
+  available_request_units,      -- FLOAT
+  refill_rate,                  -- FLOAT
+  max_request_units             -- FLOAT
+)
+```
+
+The statement completes after a new row is inserted in the system table.
+
+Updated settings are calculated based on the total consumption reported at some
+recent point in the past. The consumption value is reported as part of the
+reconfiguration request so that we can account for what happened since that
+value was reported; tokens are adjusted based on
+ - the delta between the previous consumption and the current consumption; and
+ - the refill that would have happened in the delta time period.
+
+This API should be initially called as part of creation of a new tenant (with
+`as_of_consumed_request_units = 0`). The system can start with some reasonable
+defaults (e.g. free tier settings) until that happens.
+
+In the initial implementation, this function will be called separately for each
+tenant. We will investigate batching opportunities at a later point.
+
+## Resilience considerations
+
+The system must have reasonable behavior if the bucket range becomes temporarily
+inaccessible. To achieve this, in the short term each node continues operating
+at the previous rate of consumption. Longer term, the rate can decay over time
+toward 1/N of the total refill rate, where N is the number of SQL pods.
+
+## Performance considerations
+
+The system table could get a high amount of aggregate traffic. If we have 5,000
```

Comment:
All in all, this is all scalable including within a single tenant, right? Each node only sends ~1 request per 10-30s, and that's the "atomic" unit inside of which we cannot scale. But we could put each of the 5000 nodes into their own range (hypothetically). Right?

Reply:
Yeah, even if we don't do anything special, I'd expect load-based splitting to split the table as necessary.